File: templates/problems/07_trees/15_serialize_tree/serialize_tree_test.go
package serialize_tree

import (
	"reflect"
	"testing"
)

func TestSerializeDeserialize(t *testing.T) {
	tests := []struct {
		name string
		vals []int
	}{
		{"example 1", []int{1, 2, 3, -101, -101, 4, 5}},
		{"empty tree", []int{}},
		{"single node", []int{1}},
		{"left only", []int{1, 2}},
		{"right only", []int{1, -101, 2}},
		{"full tree", []int{1, 2, 3, 4, 5, 6, 7}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			root := buildTreeHelper(tt.vals)
			data := serialize(root)
			got := treeToSlice(deserialize(data))
			want := treeToSlice(root)
			if !reflect.DeepEqual(got, want) {
				t.Errorf("serialize/deserialize roundtrip = %v, want %v", got, want)
			}
		})
	}
}

File: templates/problems/07_trees/15_serialize_tree/typescript/serialize_tree.test.ts
import { describe, it, expect } from "vitest";
import { buildTree, treeToSlice } from "./helpers";
import { serialize, deserialize } from "./serialize_tree";

describe("serialize/deserialize", () => {
    it.each([
        { name: "example 1", vals: [1, 2, 3, -101, -101, 4, 5] },
        { name: "empty tree", vals: [] },
        { name: "single node", vals: [1] },
        { name: "left only", vals: [1, 2] },
        { name: "right only", vals: [1, -101, 2] },
        { name: "full tree", vals: [1, 2, 3, 4, 5, 6, 7] },
    ])("case: $name", ({ vals }) => {
        const root = buildTree(vals);
        const data = serialize(root);
        const got = treeToSlice(deserialize(data));
        const want = treeToSlice(root);
        expect(got).toEqual(want);
    });
});
